modules/ingest-common/src/main/java/org/opensearch/ingest/common/FingerprintProcessor.java

```diff
@@ -0,0 +1,279 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.ingest.common;
+
+import org.opensearch.common.Nullable;
+import org.opensearch.common.hash.MessageDigests;
+import org.opensearch.core.common.Strings;
+import org.opensearch.ingest.AbstractProcessor;
+import org.opensearch.ingest.ConfigurationUtils;
+import org.opensearch.ingest.IngestDocument;
+import org.opensearch.ingest.Processor;
+
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.opensearch.ingest.ConfigurationUtils.newConfigurationException;
+
+/**
+ * Processor that generating hash value for the specified fields or fields not in the specified excluded list
+ */
+public final class FingerprintProcessor extends AbstractProcessor {
+    public static final String TYPE = "fingerprint";
+    // this processor is introduced in 2.16.0, we append the OpenSearch version to the hash method name to ensure
+    // that this processor always generates same hash value based on a specific hash method, if the processing logic
+    // of this processor changes in future version, the version number in the hash method should be increased correspondingly.
+    private static final Set<String> HASH_METHODS = Set.of("MD5@2.16.0", "SHA-1@2.16.0", "SHA-256@2.16.0", "SHA3-256@2.16.0");
```

Comment:
Just thinking loud about the usability here:

```
private static final Set<String> HASH_METHODS = Set.of("MD5@2.16.0", "SHA-1@2.16.0", "SHA-256@2.16.0", "SHA3-256@2.16.0");
```

This is totally fine but I believe it would be super useful to accept "MD5", "SHA-1", "SHA-256", "SHA3-256" without version suffix, meaning use the latest version of the hash. How that would work:
- we keep `HASH_METHODS` as it is designed right now, with versions
- to find the latest one, extract all supported versions of the hash method from `HASH_METHODS`, convert version part to `o.o.c.Version` and pick the latest

I think it would be the best for users, thoughts folks?

Reply:
I get the usability point, but my concern is that users have some sort of automation that creates their OpenSearch infrastructure. They define various index templates and search/ingest processor configurations etc to build new OpenSearch clusters from scratch. It might be surprising if their fingerprint processor starting changing fingerprint values for the `SHA3-256` hash method if the happened to create a cluster on an updated version. Am I overthinking this?

Reply:
>  Am I overthinking this?

It is a valid question, I think there are 2 parts to it:
 - if user cares, he has to always provide exact version
 - if user does not care, he uses what the latest is 

What I found a bit weird with **always** adding the version: for 3.0.0, my expectations would be to see / use `SHA3-256@3.0.0` but since there is no change since 2.16.0, that would not be accepted. In contrast, `SHA3-256` would clearly be accepted.

Reply:
Just as @andrross said, I have a little bit concern about that, if user upgrade an existing cluster to a new OpenSearch version which updates the fingerprint logic, then within a single index, the generated hash value for the new incoming document is different from the hash value in an existing same document, resulting in unexpected search results, users may not notice the update of the fingerprint processor, but in this case, after diving deep on that, they may fallback to set `hash_method` to an old and explicit version, and even rollback the cluster version.

Reply:
Sure, thanks @gaobinlong , it is enough evidence to discard this suggestion, thanks folks!